py/desisim/survey_release.py

```diff
@@ -0,0 +1,302 @@
+            # TODO: Implement this in desimodel instead of local path
+            pixmap=Table.read('/global/cfs/cdirs/desi/users/hiramk/desi/quickquasars/sampling_tests/npass_pixmap.fits')
+            mock_pixels = hp.ang2pix(1024, np.radians(90-mock['DEC']),np.radians(mock['RA']),nest=True)
+            try:
+                data_pixels = hp.ang2pix(1024,np.radians(90-self.data['TARGET_DEC']),np.radians(self.data['TARGET_RA']),nest=True)
+            except: 
+                data_pixels = hp.ang2pix(1024,np.radians(90-self.data['DEC']),np.radians(self.data['RA']),nest=True)
+            
+            data_passes = pixmap[data_pixels]['NPASS']
+            mock_passes = pixmap[mock_pixels]['NPASS']
+            data_pass_counts = np.bincount(data_passes,minlength=8) # Minlength = 7 passes + 1 (for bining)
+            mock_pass_counts = np.bincount(mock_passes,minlength=8)
+            mock['NPASS'] = mock_passes
+            downsampling=np.divide(data_pass_counts,mock_pass_counts,out=np.zeros(8),where=mock_pass_counts>0)
+            rand = np.random.uniform(size=len(mock))
+            if self.invert:
+                rand = 1-rand
+            selection = rand<downsampling[mock_passes]
+            log.info(f"Keeping {len(mock[selection])} mock QSOs out of {len(mock)} to match data")
+            mock = mock[selection]
+        self.mockcatalog=mock  
+
+    def assign_rband_magnitude(self,from_data=False):
+        """Assign r-band magnitudes to the catalog according to the distribution
+        
+        Args:
+            from_data (bool): if True, will use the data catalog to generate the distribution. Otherwise, will use the distribution in the default file.
+        """
+        if not from_data:
+            filename=os.path.join(os.path.dirname(desisim.__file__),'data/dn_dzdM_EDR.fits')
+            zcenters=fitsio.FITS(filename)['Z_CENTERS'][:]
+            dz = zcenters[1]-zcenters[0]
+            rmagcenters=fitsio.FITS(filename)['RMAG_CENTERS'][:]
+            dn_dzdm=fitsio.FITS(filename)['dn_dzdm'][:,:]
+            log.info(f"Generating random magnitudes according to distribution in {filename}")
+        else:
+            if self.data is None:
+                raise ValueError("No data catalog was provided")
+            dz = 0.1
+            zbins = np.arange(0,10,dz)
+            zcenters=0.5*(zbins[1:]+zbins[:-1])
+            rmagbins = np.arange(15,25,0.1)
+            rmagcenters=0.5*(rmagbins[1:]+rmagbins[:-1])
+            log.info("Generating random magnitudes according to distribution in data catalog")
+            if 'RMAG' in self.data.colnames:
+                dn_dzdm=np.histogram2d(self.data['Z'],self.data['RMAG'],bins=(zbins,rmagbins))[0]
+            elif 'FLUX_R' in self.data.colnames:
+                dn_dzdm=np.histogram2d(self.data['Z'],22.5-2.5*np.log10(self.data['FLUX_R']),bins=(zbins,rmagbins))[0]
+            else:
+                raise ValueError("No magnitude information in data catalog")
```

Comment:
Write
if 'RMAG' in self.data.colnames:
   data_rmag = self.data
elif 'FLUX_R' in self.data.colnames:
    data_rmag = 22.5-2.5*np.log10(self.data['FLUX_R'])
else:
    raise KeyError("No magnitude information in data catalog")
dn_dzdm=np.histogram2d(self.data['Z'], data_rmag, bins=(zbins,rmagbins))[0]